File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/ioc/config/BeanConfiguration.java
package de.osthus.ambeth.ioc.config;

import java.lang.reflect.Modifier;

import de.osthus.ambeth.config.IProperties;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.ioc.exception.BeanContextDeclarationException;
import de.osthus.ambeth.ioc.proxy.EmptyInterceptor;
import de.osthus.ambeth.proxy.IProxyFactory;

public class BeanConfiguration extends AbstractBeanConfiguration
{
	protected final Class<?> beanType;

	protected Object createdInstance;

	protected boolean isAbstract;

	protected final IProxyFactory proxyFactory;

	public BeanConfiguration(Class<?> beanType, String beanName, IProxyFactory proxyFactory, IProperties props)
	{
		super(beanName, props);
		this.beanType = beanType;
		this.proxyFactory = proxyFactory;
	}

	@Override
	public IBeanConfiguration template()
	{
		isAbstract = true;
		return this;
	}

	@Override
	public Class<?> getBeanType()
	{
		return beanType;
	}

	@Override
	public boolean isAbstract()
	{
		return isAbstract;
	}

	@Override
	public Object getInstance(Class<?> instanceType)
	{
		if (createdInstance == null)
		{
			try
			{
				if (instanceType.isInterface() || Modifier.isAbstract(instanceType.getModifiers()))
				{
					createdInstance = proxyFactory.createProxy(instanceType, EmptyInterceptor.INSTANCE);
				}
				else
				{
					createdInstance = instanceType.newInstance();
					if (declarationStackTrace != null && createdInstance instanceof IDeclarationStackTraceAware)
					{
						((IDeclarationStackTraceAware) createdInstance).setDeclarationStackTrace(declarationStackTrace);
					}
				}
			}
			catch (Throwable e)
			{
				if (declarationStackTrace != null)
				{
					throw new BeanContextDeclarationException(declarationStackTrace, e);
				}
				else
				{
					throw RuntimeExceptionUtil.mask(e);
				}
			}
		}
		return createdInstance;
	}
}

File: ambeth/Ambeth.IoC/ambeth/ioc/config/BeanConfiguration.cs
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Exceptions;
using De.Osthus.Ambeth.Ioc.Exceptions;
using De.Osthus.Ambeth.Ioc.Proxy;
using De.Osthus.Ambeth.Proxy;
using System;

namespace De.Osthus.Ambeth.Ioc.Config
{
    public class BeanConfiguration : AbstractBeanConfiguration
    {
        protected readonly Type beanType;

        protected readonly IProxyFactory proxyFactory;

        protected Object createdInstance;

        protected bool isAbstractConfig;

        public BeanConfiguration(Type beanType, String beanName, IProxyFactory proxyFactory, IProperties props)
            : base(beanName, props)
        {
            this.beanType = beanType;
            this.proxyFactory = proxyFactory;
        }

        public override IBeanConfiguration Template()
        {
            isAbstractConfig = true;
            return this;
        }

        public override bool IsAbstract()
        {
            return isAbstractConfig;
        }

        public override Type GetBeanType()
        {
            return beanType;
        }

        public override Object GetInstance(Type instanceType)
        {
            if (createdInstance == null)
            {
                try
                {
                    if (instanceType.IsInterface)
                    {
                        createdInstance = proxyFactory.CreateProxy(instanceType, EmptyInterceptor.INSTANCE);
                    }
                    else
                    {
                        createdInstance = Activator.CreateInstance(instanceType);
                        if (declarationStackTrace != null && createdInstance is IDeclarationStackTraceAware)
                        {
                            ((IDeclarationStackTraceAware)createdInstance).DeclarationStackTrace = declarationStackTrace;
                        }

                    }
                }
                catch (Exception e)
                {
                    if (declarationStackTrace != null)
                    {
                        throw new BeanContextDeclarationException(declarationStackTrace, e);
                    }
                    else
                    {
                        throw;
                    }
                }
            }
            return createdInstance;
        }
    }
}
